### PR TITLE
Fix json rendering of large osm ids

### LINKED ISSUE #7016: Json_render is not accurate for osm node ID
# Issue

@Rejudge-F  @SiarheiFedartsou Hi, all. We are facing an inaccurate OSM node ID format issue and I found it's related to https://github.com/Project-OSRM/osrm-backend/pull/6531.

The case is:
original node ID: 11117421192, after format: 1.111742119e+10.

The reason I think is that `PackedOSMIDs` is 34 bits and the max length of it is 11. Why do we limit the number length to less than 10? I noticed that before https://github.com/Project-OSRM/osrm-backend/pull/6531, we only limited the decimal to less than 10, but now we limit the whole number. should we rollback to the previous version or just relax the length to 11?

# Steps to reproduce
Add below test case to `unit_tests/util/json_render.cpp`.
```
    output.clear();
    renderer(11117421192);
    BOOST_CHECK_EQUAL(output, "1.111742119e+10");
```

### PR DESCRIPTION
**Pull request by imannamdari**
# Issue

Fix #7016

## Tasklist

 - [ ] CHANGELOG.md entry ([How to write a changelog entry](http://keepachangelog.com/en/1.0.0/#how))
 - [ ] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [ ] add tests (see [testing documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/testing.md))
 - [ ] review
 - [ ] adjust for comments
 - [ ] cherry pick to release branch

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: include/util/json_renderer.hpp [hunk 1]
```diff
@@ -54,7 +54,10 @@ template <typename Out> struct Renderer
         // `fmt::memory_buffer` stores first 500 bytes in the object itself(i.e. on stack in this
         // case) and then grows using heap if needed
         fmt::memory_buffer buffer;
-        fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{:.10g}"), number.value);
+        if (static_cast<std::uint64_t>(number.value) == number.value)
+            fmt::format_to(std::back_inserter(buffer), FMT_COMPILE("{}"), static_cast<std::uint64_t>(number.value));
```
[the remark this answers is not in the document]
Comment (afarber) [reply include/util/json_renderer.hpp:58]: Ah, nevermind, I see that it would break the fractional numbers

**Review by afarber — APPROVED** [review]
I have tried with a locally built and the fix works for me for example for the query below:

wget -O- 'http://127.0.0.1:5000/match/v1/driving/13.416270,52.526382;13.416849,52.526115;13.416849,52.525750;13.416538,52.525430;13.416184,52.525123;13.415755,52.525306;13.415154,52.525567;13.414596,52.525802?overview=false&generate_hints=false&gaps=ignore&annotations=nodes'

File: include/util/json_renderer.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (imannamdari) [reply include/util/json_renderer.hpp:58]: Exactly. For floating-points there gonna be problems.
The previous approach implemented before this [commit](https://github.com/Project-OSRM/osrm-backend/commit/ebd9ab4548a9959ebcf05f6bbd739bb6df8a4d1f) can be found here (and it works fine):
https://github.com/Project-OSRM/osrm-backend/blob/376282d946e0d03d9a98719d1218f54ae60e0787/include/util/json_renderer.hpp#L48

A more robust solution could be separating floats from integers. Instead of Number struct, we can have Float and Integer structs for example. Each with their own json rendering.
However, that's a significant change! :D  
If you agree with this approach (separating floats from integers), I can work on it in another pull request.

[the remark this answers is not in the document]
Comment (afarber) [reply include/util/json_renderer.hpp:58]: Dear @DennisOSRM and @SiarheiFedartsou now that CI build is working again, could you please consider merging this fix by @imannamdari  ?

In the past week I have tested his fix at my work with my matching and it really seems to fix the problem with large OSM node/way ids.

And please consider adding a new git tag `v5.28.1` after the merge?

Because the current tag `v5.27.1` is already three years old and it would be nice to mark a newer stable source code state for all users.

[the remark this answers is not in the document]
Comment (DennisOSRM) [reply include/util/json_renderer.hpp:58]: new release coming soon

[the remark this answers is not in the document]
Comment (imannamdari) [reply include/util/json_renderer.hpp]: Hi @afarber,

I've tried the example you provided, and it seems to convert the confidence value into scientific notation. The confidence value appears to be this number:
0.000000000000547

This section in json_renderer seems to be responsible for converting it to scientific notation:
FMT_COMPILE("{:.10g}")

This behavior already exists, and this merge request doesn't introduce it (since we're only addressing integer number representation in this MR).

I'm not sure what the best approach here should be. We could manually set the precision to 10 digits (similar to the [previous implementation](https://github.com/Project-OSRM/osrm-backend/blob/376282d946e0d03d9a98719d1218f54ae60e0787/include/util/json_renderer.hpp#L48)) or consider another solution. My question is: is the scientific notation really a problem? Since it's a valid representation of a float/double number, most JSON parsing frameworks should handle it without issues. Are you encountering any specific problems when parsing it?

[the remark this answers is not in the document]
Comment (afarber) [reply include/util/json_renderer.hpp]: Maybe you have different OSM data than me, I have used berlin-latest.osm.pbf

Or maybe I have misapplied your patch.

No, it parses well for me with C# System.Text.Json

[the remark this answers is not in the document]
Comment (imannamdari) [reply include/util/json_renderer.hpp]: No, I get the same result as you (5.470068842e-13). The actual double value in the code is 0.000000000000547, which is represented in JSON as 5.470068842e-13.

The issue with this number is that it is too small. The [previous implementation](https://github.com/Project-OSRM/osrm-backend/blob/376282d946e0d03d9a98719d1218f54ae60e0787/include/util/json_renderer.hpp#L48) would return 0.0000000000 in the JSON response, which was not an ideal solution for representing very small floating-point numbers. As a result, the current approach was implemented to handle the rendering of extremely small float numbers more accurately (see [this PR](https://github.com/Project-OSRM/osrm-backend/pull/6531)).

Therefore, I believe it is best to leave the current JSON rendering for floating-point numbers as it is and focus on editing it for integer numbers, which is the intent of this PR.

[the remark this answers is not in the document]
Comment (MarcelloPerathoner) [reply include/util/json_renderer.hpp]: I don't like this because it introduces two expensive casts from float to int, and it would be easier in the short term to just format everything to 11 digits.

In the medium term there should be separate json classes for doubles and integers, and maybe a precision parameter for doubles as well.

[the remark this answers is not in the document]
Comment (imannamdari) [reply include/util/json_renderer.hpp]: I agree. But we should note that these castings are already exist in the code (when we want to cast int64 to json number, which has a double type)
@DennisOSRM whats your opinion? should we temporarily change precision from 10 to 11 to solve large osm ids rendering, and then working on separating json number struct to two integer and double structs for more scalable solution?

[the remark this answers is not in the document]
Comment (MarcelloPerathoner) [reply include/util/json_renderer.hpp]: Re. separating doubles, I have following observations:

1. Variants (up to a cardinality of 11) are implemented as switch statements (at least in GNU 14.2). Adding more variants will slow the thing down a little bit.

2. The current code relies on implicit casting of numeric to double. If we add more variants we will have to sort out the implicit casts carefully.

[the remark this answers is not in the document]
Comment (MarcelloPerathoner) [reply include/util/json_renderer.hpp]: I have proposed a solution with separate treatment of OSM ids as `uint64_t` in #7142.

[the remark this answers is not in the document]
Comment (imannamdari) [reply include/util/json_renderer.hpp]: > > @DennisOSRM whats your opinion? should we temporarily change precision from 10 to 11 to solve large osm ids rendering, and then working on separating json number struct to two integer and double structs for more scalable solution?
> 
> That's a workaround that would buy some time to do the other changes I think. Any more code changes necessary to do it?

It seems that, for now, the maximum osm id is around 13,000,000,000. I increased the precision from 10 to 12, which should buy us a few more years.

After merging this, the bug should no longer appear in our upcoming release. In the meantime, we can work on @MarcelloPerathoner 's PR to implement a better solution for this issue.

[the remark this answers is not in the document]
Comment (MarcelloPerathoner) [reply include/util/json_renderer.hpp]: The problem with this patch is that it changes the format for *all* numbers. It adds even more digits where 9 digits should be amply sufficient. And this still doesn't fix the problem for people that use high IDs for private extensions.

[the remark this answers is not in the document]
Comment (imannamdari) [reply include/util/json_renderer.hpp]: f 9 digits are enough to represent a number, it will produce 9 digits (it won’t expand it to 12 digits). This is the difference between .g and .f (see the tests).

Hmm, I didn’t know about private extensions. Previously, you mentioned that as a short-term solution, instead of casting, we should simply increase the precision from 10 to 11. If this increase doesn’t solve the issue, and casting is an expensive operation (which I personally don’t believe it is, since I deployed this code in production with over 10k RPS, and the CPU usage is no different from the without-cast solution), then we don’t have a viable short-term solution.

[the remark this answers is not in the document]
Comment (MarcelloPerathoner) [reply include/util/json_renderer.hpp]: Private extensions are mentioned here: #7069

[the remark this answers is not in the document]
Comment (plecompt446) [reply include/util/json_renderer.hpp]: Is this change incoorperated in v6.0.0 because it still has this problem on this day? When I use v5.25.0 I do not have this issue, but v6.0.0 is significantly faster which I would prefer to use that one if possible

**Comment by fenwuyaoji** [comment]
已收到您的邮件，我将及时查看并回复，谢谢                                                                                                                     王鑫

File: include/util/json_renderer.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (plecompt446) [reply include/util/json_renderer.hpp]: or is there an alternative docker image that I can pull where this specific issue is resolved as today I still get the id's in scientific notation when using v6.0.0?

**Comment by kevinburke** [comment]
Yes it is still relevant

**Comment by afarber** [comment]
This issue has been open for over a year now, and it affects anyone using OSRM with recent OSM data. For context, OSM node IDs exceeded 10 billion in August 2025, and this threshold continues to grow.

Other routing projects have already addressed this:

- Valhalla: Fixed in https://github.com/valhalla/valhalla/pull/2422 (2019)
- Osmosis: Supported since v0.42 (Feb 2013)
- Overpass API: Full support since v0.7.53 (Sep 2016)

I would like to submit my own PR with a more robust fix that includes:

- Integer detection to avoid scientific notation for large OSM IDs
- Graceful handling of NaN/Infinity values
- Additional unit tests covering edge cases

I hope a fresh PR with additional tests might help move this forward.

**Comment by DennisOSRM** [comment]
Please submit your PR

**Comment by TheMarex** [comment]
Thanks your contribution @imannamdari and thanks to @afarber for getting the fix out. Closing here.
